app/post_persistent_notification.go

```diff
@@ -0,0 +1,380 @@
+		return model.NewAppError("DeletePersistentNotificationsPost", "app.post_priority.delete_persistent_notification_post.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
+	} else if len(posts) == 0 {
+		// Either the notification post already deleted or was never a notification post
+		return nil
+	}
+
+	if !*a.Config().ServiceSettings.AllowPersistentNotificationsForGuests {
+		user, nErr := a.Srv().Store().User().Get(context.Background(), c.Session().UserId)
+		if nErr != nil {
+			var nfErr *store.ErrNotFound
+			switch {
+			case errors.As(nErr, &nfErr):
+				return model.NewAppError("DeletePersistentNotificationsPost", MissingAccountError, nil, "", http.StatusNotFound).Wrap(nErr)
+			default:
+				return model.NewAppError("DeletePersistentNotificationsPost", "app.user.get.app_error", nil, "", http.StatusInternalServerError).Wrap(nErr)
+			}
+		}
+		if user.IsGuest() {
+			c.Logger().Debug("DeletePersistentNotificationsPost: Persistent Notification feature is not enabled for guests.")
+			return nil
+		}
+	}
+
+	// if checkMentionedUser is "false" that would mean user is already authorized to delete the persistent-notification.
+	isUserMentioned := !checkMentionedUser
+
+	// post owner is not allowed to stop the persistent notifications via ack, reply or reaction.
+	if checkMentionedUser && mentionedUserID != post.UserId {
+		if err := a.forEachPersistentNotificationPost([]*model.Post{post}, func(_ *model.Post, _ *model.Channel, _ *model.Team, mentions *ExplicitMentions, _ model.UserMap, _ map[string]map[string]model.StringMap) error {
+			if mentions.isUserMentioned(mentionedUserID) {
+				isUserMentioned = true
+			}
+			return nil
+		}); err != nil {
+			return model.NewAppError("DeletePersistentNotificationsPost", "app.post_priority.delete_persistent_notification_post.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
+		}
+	}
+
+	if isUserMentioned {
+		if err := a.Srv().Store().PostPersistentNotification().Delete([]string{post.Id}); err != nil {
+			return model.NewAppError("DeletePersistentNotificationsPost", "app.post_priority.delete_persistent_notification_post.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
+		}
+	}
+
+	return nil
+}
+
+func (a *App) SendPersistentNotifications() error {
+	notificationInterval := time.Duration(*a.Config().ServiceSettings.PersistentNotificationInterval) * time.Minute
+	notificationMaxCount := int16(*a.Config().ServiceSettings.PersistentNotificationMaxCount)
+
+	// fetch posts for which first notificationInterval duration has passed
+	maxCreateAt := time.Now().Add(-notificationInterval).UnixMilli()
+
+	pagination := model.CursorPagination{
+		Direction: "down",
+		PerPage:   500,
+	}
+
+	// Pagination loop
+	for {
+		notificationPosts, hasNext, err := a.Srv().Store().PostPersistentNotification().Get(model.GetPersistentNotificationsPostsParams{
+			MaxCreateAt:   maxCreateAt,
+			MaxLastSentAt: maxCreateAt,
+			Pagination:    pagination,
+		})
```

Comment:
I'm not sure, what do you mean by _richer SQL query_. There are certain conditions which need to be met, to decide which PN posts are fetched.

Reply:
So one idea would be pass `notificationInterval` into `Get()` and have the database decide which posts to return. Since we immediately process them and update the records, the next query should automatically return the next set of posts without needing to maintain a cursor, and we can just exit when the set is empty.

Reply:
> Would it make sense to introduce a richer SQL query that just gets notifications ready to be notified? In that case, we wouldn't need complex pagination, as we could literally just get the next 1000 in the order returned, act on them, and then repeat until the list is empty.

@lieut-data if we don't paginate won't we get the same posts over and over again, until they expire? or am I missing something?

e.g:

We have 50 PersistentNotification posts that are not deleted.
20 have been sent 5 times, 20 have been sent 3 times, and 10 have been sent 0 times.
We need to fire 1 notification for every one of them.
Let's say we limit the query at 20.

At the fist iteration the 0-20 will run and then marked to be deleted.
At the second iteration 20-40 will run for the 4th time in their life,
At the third iteration we'll get again the 20-40 and they'll run for the 5th time
etc...

Reply:
@lieut-data I've updated the logic. Now `Get` returns next set of valid posts to be notified and `DeleteExpired` takes care of deleting the invalid posts.